### PR TITLE
Fast score matrix

### PR DESCRIPTION
**Pull request by RK900**

### CODE REVIEW

File: epitome/functions.py [hunk 1]
```diff
@@ -179,6 +179,93 @@ def indices_for_weighted_resample(data, n,  matrix, cellmap, assaymap, weights =
     surrounding = np.unique(list(map(func_, choice)))
     return surrounding[(surrounding > 0) & (surrounding < data_count)].astype(int)
 
+def compute_casv(m1, m2, radii, indices= None):
+    '''
+    Computes CASV between two matrices. CASV indiciates how similar
```
Comment (akmorrow13) [inline epitome/functions.py:184]: https://github.com/YosefLab/epitome/blob/master/epitome/generators.py#L216 see code here to add a continous bool parameter to see whether user wants to include agreement

File: epitome/functions.py [hunk 2]
```diff
@@ -179,6 +179,93 @@ def indices_for_weighted_resample(data, n,  matrix, cellmap, assaymap, weights =
...
+            radius_indices = np.concatenate(radius_ranges)
+
+            # data from known cell types (m1 portion)
+            m1_slice = m1[radius_indices, :]
+            m2_slice = np.repeat(m2[radius_indices, :, :],axis=1, repeats = ncells)
+            
+
+            # shape: radius size x (nassaysxncells) by nsamples
+            pos = (m1_slice.T*m2_slice.T).T
+#             agree = (m1_slice.T == m2_slice.T).T
```
Comment (akmorrow13) [inline epitome/functions.py:239]: can uncomment these and incorporate agree on line 253

File: epitome/generators.py [hunk 3]
```diff
@@ -199,14 +199,14 @@ def g():
...
-                        agree = cell_train_data == similarity_matrix[:,radius_indices]
+                        # agree = cell_train_data == similarity_matrix[:,radius_indices]
 
                     else:
                         cell_label_data = data[label_cell_indices[delete_indices][:,None],radius_indices]
 
                         # remove middle dimension and flatten similarity targets
                         pos = (cell_train_data*cell_label_data)
-                        agree = (cell_train_data == cell_label_data)
+                        # agree = (cell_train_data == cell_label_data)
```
Comment (akmorrow13) [inline epitome/generators.py:209]: can add back in

File: epitome/models.py [hunk 4]
```diff
@@ -746,3 +750,149 @@ def create_model(self, **kwargs):
         model = tf.keras.models.Model(inputs=cell_inputs, outputs=outputs)
 
         return model
+    
+    def score_matrix_fast(self, accessibility_peak_matrix, regions):
+        """ Runs predictions on a matrix of accessibility peaks, where columns are samples and
+        rows are regions from regions_peak_file. rows in accessilibility_peak_matrix should matching
```
Comment (akmorrow13) [inline epitome/models.py:756]: this can replace the score_matrix function

File: epitome/models.py [hunk 5]
```diff
@@ -27,7 +27,9 @@
 import tqdm
 import logging
 import sys
-import os
+import copy
```
Comment (akmorrow13) [inline epitome/models.py:30]: where is this used?

**Review by akmorrow13 — CHANGES_REQUESTED** [review]

File: epitome/models.py [hunk 6]
```diff
@@ -746,3 +750,149 @@ def create_model(self, **kwargs):
...
+                for celltype in range(num_celltypes):
+                    idx = (num_targets + 4) * celltype
+                    if idx >= len(selected_gen[0]):
+                        break
+                    num_targets = first_substring(naming_scheme[idx:idx+len_feats_per_celltype+total_targets], '_agree', naming_scheme[idx-10:idx+len_feats_per_celltype+10], (idx, len_feats_per_celltype))
+                    casv_cell = selected_casv[:, celltype]
+                    selected_gen[0][idx + num_targets : idx + num_targets + 4] = casv_cell
+        
+
+        results = []
```
Comment (akmorrow13) [inline epitome/models.py:879]: instead of returning a matrix, we can have a parameter that specifies where to write an h5 file. then we can iteratively write each slice to the h5 file

File: epitome/models.py [hunk 7]
```diff
@@ -746,3 +750,149 @@ def create_model(self, **kwargs):
...
+        for c in tqdm(range(num_cells)):
+            for r in range(num_regions):
+                results.append(self._predict(to_stack[c, r, :][0][None, :]))
```
Comment (akmorrow13) [inline epitome/models.py:882]: We should pass in all the regions at once per cell

File: epitome/test/generators_test.py [hunk 8]
```diff
@@ -12,8 +12,12 @@ def __init__(self, *args, **kwargs):
 		# get shape of train data
 		eligible_cells = ['K562','HepG2','H1','A549','HeLa-S3']
 		eligible_targets = ['DNase','CTCF']
-		matrix, cellmap, targetmap = self.getFeatureData(eligible_targets, eligible_cells)
-		self.train_shape = (np.max(matrix)+1, 1800)  # 1800 is size of toy data training
+		self.matrix, self.cellmap, self.targetmap = self.getFeatureData(eligible_targets, eligible_cells)
+		self.train_shape = (np.max(self.matrix)+1, 1800)  # 1800 is size of toy data training
+		self.data = np.zeros(self.train_shape)
+		self.data[::2] = 1
```
Comment (akmorrow13) [inline epitome/test/generators_test.py:18]: can you push unit tests so we can take a look?

File: epitome/functions.py (hunk 1 above)
Comment (RK900) [reply epitome/functions.py:184]: is this for `compute_casv` or for `load_data`?

File: epitome/models.py [hunk 9]
```diff
@@ -746,3 +751,148 @@ def create_model(self, **kwargs):
...
+                 mode = Dataset.RUNTIME,
+                 similarity_matrix = matrix,
+                 similarity_targets = ['DNase'],
+                 indices = indices,
+                 continuous=True,
+                 return_feature_names=True)
+
+        gen_to_list = list(gen())
+        to_stack = list(to_stack())
+        gen_to_list = np.array(gen_to_list)
```
Comment (akmorrow13) [inline epitome/models.py:805]: doesn't this still have the labels and masks in it? so each item in the array is a tuple? how does this work when you pass it into the CASV?

File: epitome/models.py [hunk 10]
```diff
@@ -746,3 +751,148 @@ def create_model(self, **kwargs):
...
+                old_i = i
+            indices_to_merge.append((old_idx, len(conversionObject.joined.idx)))
+            
+            a = np.stack(added_indices)
+        else:
+            a = np.transpose(accessibility_peak_matrix, axes=[1, 0])
+        
+        a = a[:, None, :]
+        
+        out = compute_casv(gen_to_list, a, radii)
```
Comment (akmorrow13) [inline epitome/models.py:838]: Upon looking at the new score_matrix function, I believe it is exhibiting incorrect behavior, and compute_casv is not the issue (although it should be able to handle a division by 0 error). The issue is that you are passing in a 4 x 1 x 2 matrix for m2, and similarly shaped for m1. However, the region dimensions should be the full size of the dataset (in the test case, 2100 regions). This correct behavior can be seen in the original score_matrix function, where you pass in the entire matrices, then use the indices parameter (now in compute_casv) to specify the indices you want to compute on

**Review by akmorrow13 — CHANGES_REQUESTED** [review]

File: epitome/models.py (hunk 9 above)
Comment (RK900) [reply epitome/models.py:805]: `to_stack` isn't passed into the CASV. I added a parameter in `load_data` (`keep_lbl_mask`) which if set to True, returns the labels and masks. For `gen_to_list`, this flag is set to false so it doesn't have the labels or masks.
